userspace/libsinsp/sinsp.cpp

```diff
@@ -662,22 +700,46 @@ void sinsp::open_savefile(const std::string& filename, int fd)
 	params.start_offset = 0;
 	params.fbuffer_size = 0;
 	oargs.engine_params = &params;
-	open_common(&oargs);
+
+	// AFAICT this is because tinfo==NULL when calling the callback in scap_read_fdlist
+	struct scap_platform* platform = scap_savefile_alloc_platform(nullptr, nullptr);
+	open_common(&oargs, &scap_savefile_engine, platform);
+#else
+	throw sinsp_exception("SAVEFILE engine is not supported in this build");
+#endif
 }
 
 void sinsp::open_plugin(const std::string& plugin_name, const std::string& plugin_open_params, scap_mode_t mode)
 {
+#ifdef HAS_ENGINE_SOURCE_PLUGIN
 	scap_open_args oargs = factory_open_args(SOURCE_PLUGIN_ENGINE, mode);
 	struct scap_source_plugin_engine_params params;
 	set_input_plugin(plugin_name, plugin_open_params);
 	params.input_plugin = &m_input_plugin->as_scap_source();
 	params.input_plugin_params = (char*)m_input_plugin_open_params.c_str();
 	oargs.engine_params = &params;
-	open_common(&oargs);
+
+	struct scap_platform* platform;
+	switch(mode)
+	{
+		case SCAP_MODE_PLUGIN:
+			platform = scap_generic_alloc_platform(::on_new_entry_from_proc, this);
```

Comment:
Engine vtables are const and I don't think I'm ready for mutable ones ;)

Reply:
I removed `mode` from `scap_open_args` and pass an optional override as an extra param to sinsp::open_common. I'm not 100% a fan of this tbh but removing the mode from scap_open_args feels worth it.

Reply:
ok seems a reasonable solution, thank you! The only thing that i don't love is that the v-table could bring a mode that is different to the real one with which we opened the engine...Moreover, probably, if we override the v-table we can remove the `m_mode` flag in the inspector and use only the engine v-table content...but yes we lose the const qualifier

Reply:
I don't love (to put it mildly :)) the concept of *a* mode in the first place. Just as in libscap, it's static polymorphic dispatch in disguise and right now it has two components (engine/platform) to abstract over. IMO the right way to handle m_mode is to avoid breaking it until we can finally get rid of it by making the polymorphism explicit.

Now that I think about it, the only user of scap_vtable->mode is sinsp, so we should drop the mode from the scap vtable instead.

Reply:
Yeah that's a good point, probably the best thing is to remove it from the v-table, BTW i won't block this PR we can also do this in a follow-up

Reply:
Give me a minute please :)

Reply:
well okay, 10 minutes :)

Reply:
Amazing thank you!